pkg/templates/kubeadm/v1beta4/kubeadm.go

```diff
@@ -0,0 +1,621 @@
+	nodeRegistration.IgnorePreflightErrors = []string{
+		"DirAvailable--etc-kubernetes-manifests",
+	}
+
+	controlPlaneEndpoint := fmt.Sprintf("%s:%d", cluster.APIEndpoint.Host, cluster.APIEndpoint.Port)
+
+	joinConfig := &kubeadmv1beta4.JoinConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "kubeadm.k8s.io/v1beta4",
+			Kind:       "JoinConfiguration",
+		},
+		Discovery: kubeadmv1beta4.Discovery{
+			BootstrapToken: &kubeadmv1beta4.BootstrapTokenDiscovery{
+				Token:                    s.JoinToken,
+				APIServerEndpoint:        controlPlaneEndpoint,
+				UnsafeSkipCAVerification: true,
+			},
+		},
+	}
+
+	if cluster.CloudProvider.External {
+		nodeRegistration.KubeletExtraArgs = setAllArgsValue(nodeRegistration.KubeletExtraArgs, "cloud-provider", "external")
+	}
+
+	joinConfig.NodeRegistration = nodeRegistration
+
+	return &Config{
+		JoinConfiguration: joinConfig,
+	}, nil
+}
+
+func newInitConfiguration(bootstrapToken *bootstraptokenv1.BootstrapTokenString, advertiseAddress string) *kubeadmv1beta4.InitConfiguration {
+	return &kubeadmv1beta4.InitConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "kubeadm.k8s.io/v1beta4",
+			Kind:       "InitConfiguration",
+		},
+		BootstrapTokens: []bootstraptokenv1.BootstrapToken{
+			{
+				Token: bootstrapToken,
+				Groups: []string{
+					"system:bootstrappers:kubeadm:default-node-token",
+				},
+				TTL: &metav1.Duration{
+					Duration: bootstrapTokenTTL,
+				},
+				Usages: []string{
+					"signing",
+					"authentication",
+				},
+			},
+		},
+		LocalAPIEndpoint: kubeadmv1beta4.APIEndpoint{
+			AdvertiseAddress: advertiseAddress,
+		},
+	}
+}
+
+func newJoinConfiguration(advertiseAddress, joinToken, controlPlaneEndpoint string) *kubeadmv1beta4.JoinConfiguration {
+	return &kubeadmv1beta4.JoinConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "kubeadm.k8s.io/v1beta4",
+			Kind:       "JoinConfiguration",
+		},
+		ControlPlane: &kubeadmv1beta4.JoinControlPlane{
+			LocalAPIEndpoint: kubeadmv1beta4.APIEndpoint{
+				AdvertiseAddress: advertiseAddress,
+			},
+		},
+		Discovery: kubeadmv1beta4.Discovery{
+			BootstrapToken: &kubeadmv1beta4.BootstrapTokenDiscovery{
+				Token:                    joinToken,
+				APIServerEndpoint:        controlPlaneEndpoint,
+				UnsafeSkipCAVerification: true,
+			},
+		},
+	}
+}
+
+func newNodeRegistration(s *state.State, host kubeoneapi.HostConfig) kubeadmv1beta4.NodeRegistrationOptions {
+	kubeletCLIFlags := []kubeadmv1beta4.Arg{
+		{
+			Name:  "volume-plugin-dir",
+			Value: "/var/lib/kubelet/volumeplugins",
+		},
+	}
+	if s.Cluster.CloudProvider.External {
+		kubeletCLIFlags = append(kubeletCLIFlags, kubeadmv1beta4.Arg{
+			Name:  "cloud-provider",
+			Value: "external",
+		})
+	}
+
+	// If external or in-tree CCM is in use we don't need to set --node-ip
+	// as the cloud provider will know what IPs to return.
+	if s.Cluster.ClusterNetwork.IPFamily.IsDualstack() {
+		if !s.Cluster.CloudProvider.External {
```

Comment:
The comment seemingly doesn't match the code. It looks like if it's not dualstack, we _always_ set --node-ip, and even if it's dualstack, we still set if for in-tree CCMs?

Reply:
Exactly, the comment is now updated.